### PR TITLE
Remove groupby copy

### LINKED ISSUE #2753: Remove `groupby.copy`
We have previously added the `copy` method to `groupby`. This allows for extracting a group with a specific coord value. It is also used for the implementation of `sort` and boolean-indexing. #2745 highlights an implementation issue, as well as a number of performance issues. In addition to that, the interface is quite error prone. Consider:

```python
da.groupby('label').copy(0)
```

It is totally unpredictable what this will do, since we do not know which values the `'label'` coord contains.

Since boolean indexing can do the job in most cases in a more expressive and safer manner, we should encourage to use this instead:

```python
da[da.coords['label'] == 1]
da[da.coords['x'] == sc.scalar(1.2, unit='m')]
```

If we succeed to replace the implementation of `sort` and boolean indexing by `bin` then I think we should therefore remove the `copy` method from `groupby`.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2753.

"Side" effect: `sc.sort` is now much faster:

```python
import scipp as sc
da = sc.data.table_xyz(10_000_000)
sc.sort(da, 'x')
```

`sc.sort` timings before: 16 seconds, after 2 seconds

### CODE REVIEW

File: lib/dataset/sort.cpp [hunk 1]
```diff
@@ -3,45 +3,83 @@
...
+        return nan_sensitive_less(b.first, a.first);
+      });
+
+    auto indices =
+        makeVariable<scipp::index_pair>(Dims{key.dim()}, Shape{size});
+    scipp::index j = 0;
+    for (auto &index : indices.values<scipp::index_pair>().as_span()) {
+      const auto start = pairs[j++].second;
+      index = {start, start + 1};
+    }
```
Comment (jl-wynen) [inline lib/dataset/sort.cpp:45]: Can this be done with `std::transform`? Took me a moment to understand the loop.

File: lib/dataset/sort.cpp [hunk 2]
```diff
@@ -3,45 +3,83 @@
...
+    if (std::isnan(b))
+      return !std::isnan(a);
+  return a < b;
+};
+
+template <class T> struct IndicesForSorting {
+  static Variable apply(const Variable &key, const SortOrder order) {
+    const auto size = key.dims()[key.dim()];
+    const auto values = key.values<T>();
+    std::vector<std::pair<T, scipp::index>> pairs;
```
Comment (jl-wynen) [inline lib/dataset/sort.cpp:24]: 'pairs' is too generic. What about 'key_index' or 'key_index_pairs'?

File: lib/dataset/sort.cpp [hunk 3]
```diff
@@ -3,45 +3,83 @@
...
+    pairs.reserve(size);
+    scipp::index i = 0;
+    for (const auto &value : key.values<T>())
+      pairs.emplace_back(value, i++);
```
Comment (jl-wynen) [inline lib/dataset/sort.cpp:28]: Maybe but this into a separate function. It leaks `i` and I had to check that it is not used anywhere below.

**Review by jl-wynen — COMMENTED** [review]
Just some readability comments. Looks good otherwise.

File: lib/dataset/sort.cpp [hunk 4]
```diff
@@ -21,28 +21,33 @@ template <class T> struct IndicesForSorting {
...
 
     auto indices =
         makeVariable<scipp::index_pair>(Dims{key.dim()}, Shape{size});
-    scipp::index j = 0;
-    for (auto &index : indices.values<scipp::index_pair>().as_span()) {
-      const auto start = pairs[j++].second;
-      index = {start, start + 1};
-    }
+    std::transform(key_index.begin(), key_index.end(),
+                   indices.values<scipp::index_pair>().as_span().begin(),
```
Comment (jl-wynen) [inline lib/dataset/sort.cpp:47]:
```suggestion
                   indices.values<scipp::index_pair>().begin(),
```
?

Comment (SimonHeybrock) [reply lib/dataset/sort.cpp:47]: The `as_span` can avoid a potentially more expensive iteration (using `MultiIndex`).

Comment (jl-wynen) [reply lib/dataset/sort.cpp:47]: ok

**Review by jl-wynen — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
Using `tbb::parallel_sort` gives a further speedup of 1.8 seconds -> 1.1 seconds.

**Pull request merged**